Configuration Overhaul RFC

0x009922 opened the pull request:
[Rendered](https://github.com/0x009922/iroha-rfcs/blob/config-overhaul/000-config-overhaul.md)

---

Original PR:

- https://github.com/hyperledger/iroha/pull/3795

### Checklist

- [ ] Tech writers review
- [x] Make a note about migration ([context](https://github.com/hyperledger/iroha/pull/3795#discussion_r1319702299))
- [x] Address handling of dev-related fields ([context](https://github.com/hyperledger/iroha-rfcs/pull/3#issuecomment-1715540646))

0x009922 committed:
docs: migrate document from hyperledger/iroha#3795

Signed-off-by: Dmitry Balashov <a.marcius26@gmail.com>

mversic commented:
I think we should also add a way to inject __private__ parameters in a configuration. Those parameters should start with an underscore and should never be a part of our public API contract (and would most likely be only used during testing). If any user depends on this parameter it's on their own responsibility.

A good example of such a parameter is `sumeragi::debug_force_soft_fork` which, according to what I wrote, should be `sumeragi::_force_soft_fork`

0x009922 commented:
@mversic, I thought about this topic too when faced those several parameters "for development/debugging/testing purposes". Agreed that they should not be a part of our public API.

As for the format, I have some alternatives. IMO, using named nested sections for it might be more explicit:

```toml
[sumeragi.debug]
force_soft_fork = true

[kura.debug]
output_new_blocks = true
```

Another way might be to have an entirely separate `debug` section with all of the parameters:

```toml
[debug]
sumeragi.force_soft_fork = true
kura.output_new_blocks = true
```

Not sure what keyword to use for it: `dev`, `debug`, `test`, `unstable`, or something else?

0x009922 committed:
feat: apply minor changes

- Include dev-only fields
- Include migration tool mention
- Include human-readable types proposal
- Remove "dual view" proposal as extra

Signed-off-by: Dmitry Balashov <a.marcius26@gmail.com>

0x009922 commented:
New changes:

- Combined a proposal "reconsider parameter naming conventions", which includes snake_case, dev fields ("private" fields), and more renaming samples.
- Added a brief proposal for human-readable types
- Removed "dual view" proposal
- In the implementation plan, include a migration tool development

0x009922 committed:
chore: assign `001` number

Signed-off-by: Dmitry Balashov <a.marcius26@gmail.com>
